Improve error handling for routing calls

gmellemstrand opened the pull request:
To be completed by pull request submitter:

- [ ] **issue**: Link to or create an [issue](https://github.com/opentripplanner/OpenTripPlanner/issues) that describes the relevant feature or bug. Add [GitHub keywords](https://help.github.com/articles/closing-issues-using-keywords/) to this PR's description (e.g., `closes #45`).
- [ ] **roadmap**: Check the [roadmap](https://github.com/orgs/opentripplanner/projects/1) for this feature or bug. If it is not already on the roadmap, PLC will discuss as part of the review process.
- [ ] **tests**: Have you added relevant test coverage? Are all the tests passing on [the continuous integration service (Travis CI)](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#continuous-integration)?
- [ ] **formatting**: Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#code-style)? 
- [ ] **documentation**: If you are adding a new configuration option, have you added an explanation to the [configuration documentation](docs/Configuration.md) tables and sections?
- [ ] **changelog**: add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue

To be completed by @opentripplanner/plc:

- [ ] reviews and approvals by 2 members, ideally from different organizations
- [ ] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

This pull request further separates the internal error handling from the rest and Transmodel APIs. It also allows catches exceptions from transit routing and direct street routing separately, allowing one to fail while the other returns a result.

What I have done:
- Move Request/Response classes to routing/api package.
- Created a new RoutingError class, as well as RoutingErrorCode and InputField enums.
- Created a new RoutingValidationException which contains a RoutingError.
- Catch RoutingValidationExceptions from transit routing and direct street routing separately. A list of RoutingErrors is returned with the RoutingResponse. Other exceptions are not caught at this level.
- Differentiate between LOCATION_NOT_FOUND and NO_STOPS_IN_RANGE. OUTSIDE_BOUNDS and OUTSIDE_SERVICE_PERIOD still not in use.
- Created PlannerErrorMapper class to to map RoutingError to REST API PlannerError.

t2gran commented on src/main/java/org/opentripplanner/routing/api/request/RoutingService.java:
```diff
@@ -1,4 +1,4 @@
-package org.opentripplanner.routing;
+package org.opentripplanner.routing.api.request;
```
The RoutingService is the main entry point of the `org.opentripplanner.routing` package and should be in the root.

t2gran commented on src/main/java/org/opentripplanner/routing/api/response/RoutingErrorCode.java:
```diff
@@ -0,0 +1,25 @@
+package org.opentripplanner.routing.api.response;
+
+public enum RoutingErrorCode {
+  SYSTEM_ERROR,
```
We should throw OTPApplicationException to exit a search, so we do not need the SYSTEM_ERROR here. we do need it on the API enum.

t2gran commented on src/main/java/org/opentripplanner/routing/api/response/RoutingErrorCode.java:
```diff
@@ -0,0 +1,25 @@
+   */
+  OUTSIDE_SERVICE_PERIOD,
+
+  /**
+   * The location could not be matched to a valid id or coordinate.
+   */
+  LOCATION_NOT_FOUND,
```
I think we need to be more specific: FROM_PLACE_NOT_FOUND, TO_PLACE_NOT_FOUND (INTERMEDIATE_PLACE_NOT_FOUND when supporting via search)

gmellemstrand replied:
The idea here was to use the paramValue in RoutingError to specify which place was not found.

t2gran replied:
Maybe it is better to add another enum to report back the incorrect input field. Then that could also be used with the next error code.

gmellemstrand replied:
Changed the paramValue to an enum, and then it is checked in the PlannerErrorMapper and mapped to the correct Message.

t2gran commented on src/main/java/org/opentripplanner/routing/api/response/RoutingErrorCode.java:
```diff
@@ -0,0 +1,25 @@
+   */
+  LOCATION_NOT_FOUND,
+
+  /**
+   * The location was found, but no stops could be connected within the search radius.
+   */
```
Improve doc: "Location found, but no stops found within the search radius."

t2gran commented on src/main/java/org/opentripplanner/routing/api/response/RoutingError.java:
```diff
@@ -0,0 +1,13 @@
+public class RoutingError {
+  public final RoutingErrorCode code;
+  public final String message;
+  public final String paramValue;
```
The parameter value here is probably not a good idea, I know I suggested it, but since we want typed parameters as input, the "original" string is not available. So the best we can do is to report back witch input field that failed to validate. We should use an enum for the input fields.

gmellemstrand replied:
I have changed it to an enum

t2gran commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -63,29 +65,53 @@ public RoutingWorker(RaptorConfig<TripSchedule> config, RoutingRequest request)
             // Direct street routing
-            itineraries = new ArrayList<>(DirectStreetRouter.route(router, request));
+            itineraries.addAll(DirectStreetRouter.route(router, request));
+        } catch (Exception e) {
```
You should only catch business specific exceptions that can be mapped to a RoutingError.

t2gran commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -63,29 +65,53 @@ public RoutingWorker(RaptorConfig<TripSchedule> config, RoutingRequest request)
     public RoutingResponse route(Router router) {
-        try {
-            List<Itinerary> itineraries;
+        List<Itinerary> itineraries = new ArrayList<>();
+        List<RoutingError> routingErrors = new ArrayList<>();
```
I would make the routingErrors a member of the request scoped Worker class - then each of the two try/catch blocks can be moved into their own methods and don´t interrupt the reading of the main flow. This have the side-effect of changing the state of the worker, so that should be documented for each of the methods.

gmellemstrand replied:
As you suggested in another comment, I've instead opted to use a general RoutingValidationException that includes a RoutingError.

t2gran commented on src/main/java/org/opentripplanner/routing/error/NoStopsInRangeException.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Indicates that a vertex was found, but no stops could be found within the search radius.
+ */
+public class NoStopsInRangeException extends RuntimeException {
+  private static final long serialVersionUID = 1L;
+
+  List<String> missing;
+
+  public NoStopsInRangeException(List<String> missing) {
+    super("vertices not found: " + missing.toString());
+    this.missing = missing;
+  }
+
+  public List<String> getMissing() {
+    return missing;
+  }
+}
```
We do not need to distinguish between different error situation, so this exception should be more general like e.g. `RoutingValidationException` and include the RoutingError. Then maybe the error handling in RoutingWorker collapses to one line and you do not need to make the `routingErrors` class scoped (witch is not ideal).

gmellemstrand replied:
I have refactored the code to use the RoutingValidationException with a RoutingError included. This also meant I could delete som of the other Exception types.

t2gran commented on src/main/java/org/opentripplanner/routing/api/response/RoutingErrorCode.java:
```diff
@@ -0,0 +1,23 @@
+package org.opentripplanner.routing.api.response;
+
+public enum RoutingErrorCode {
+  /**
+   * Location was found, but it was located outside the street network.
+   */
+  OUTSIDE_BOUNDS,
+
+  /**
+   * The time provided was outside the available service period.
+   */
+  OUTSIDE_SERVICE_PERIOD,
+
+  /**
+   * The location could not be matched to a valid id or coordinate.
+   */
+  LOCATION_NOT_FOUND,
+
+  /**
+   * The location was found, but no stops could be found within the search radius.
+   */
+  NO_STOPS_IN_RANGE
+}
```
Add some TODO OTP2 here for unused codes.

t2gran commented:
Take a look at the bundled client and see if it works with this PR.

gmellemstrand commented:
When testing with the built-in client, I discovered that the LOCATION_NOT_FOUND error is not caught in the right place. It seems like each of the APIs are setting the RoutingContext outside of the call to the RoutingService. I have now moved setting the RoutingContext inside the direct and transit routing methods, in order for the exceptions to be caught the same way as other routing errors. Each of those methods had a call to the checkIfVerticesFound method. I have removed those calls, as it is already done within the setRoutingContext method.
